### ISSUE TITLE
Issue with Capturing Log Scope in Application Insights with Azure Function

### ISSUE THREAD

**Issue by CodeRedPaul** [microsoft/ApplicationInsights-dotnet#2089]
Hi there

I have a Service Bus trigger on an Azure Function and am having some issues capturing the defined scope data in Application Insights.

I am using AutoFac for DI / as an IoC container and am setting things up so that I can manually configure items such as Sampling and TelemetryChannel etc (not all done yet).

I am using v 6.0.0.53 of the "Autofac.Extensions.DependencyInjection.AzureFunctions" NuGet package to assist in wiring up AutoFac and am using v.1.1.0 of Microsoft.Azure.Functions.Extensions and v. 2.15.0 of Microsoft.ApplicationInsights.

The issue is that although I am successfully getting Trace results in Application Insights I do not appear to be receiving the scope details. It will be massively helpful to my project to use Structured Logging so I'd really like to work out what the issue is,

Firstly, here is an excerpt of my services configuration using AutoFac:

```
builder.RegisterType<ServerTelemetryChannel>().AsSelf().SingleInstance();
builder.Register(ctx => ctx.Resolve<ServerTelemetryChannel>()).As<ITelemetryChannel>().SingleInstance();

builder.Register(
	ctx =>
	{
		var appConfig = ctx.Resolve<IConfiguration<ConfigurationKeys>>();
		bool LevelFunc(LogLevel level) => level >= Enums.Parse<LogLevel>(appConfig.GetEntry(ConfigurationKeys.LogLevel), true);
		var appInsightsInstrumentationKey = appConfig.GetEntry(ConfigurationKeys.AppInsightsInstrumentationKey);
		return new ServiceCollection()
			.AddLogging(loggingBuilder => loggingBuilder.AddConsole())
			.AddLogging(loggingBuilder =>
			{
				loggingBuilder.AddProvider(
					new ApplicationInsightsLoggerProvider(
						new OptionsWrapper<TelemetryConfiguration>(
							new TelemetryConfiguration(appInsightsInstrumentationKey)),
						new OptionsWrapper<ApplicationInsightsLoggerOptions>(
							new ApplicationInsightsLoggerOptions { IncludeScopes = true })
					)
				);
			})
			.AddLogging(loggingBuilder => loggingBuilder.AddFilter(LevelFunc))
			.AddLogging(loggingBuilder => loggingBuilder.AddFilter<ApplicationInsightsLoggerProvider>(LevelFunc))
			.Configure<TelemetryConfiguration>(
				config =>
				{
					var serverChannel = ctx.Resolve<ServerTelemetryChannel>();
					config.TelemetryChannel = serverChannel;
					serverChannel.Initialize(config);
				}
			)
			.BuildServiceProvider()
			.GetService<ILoggerFactory>();
	})
	.As<ILoggerFactory>()
	.SingleInstance()
	;
builder.Register(
		ctx => ctx.Resolve<ILoggerFactory>().CreateLogger<MockServiceListener>()
	)
	.As<ILogger<MockServiceListener>>()
	.SingleInstance()
	;
```

You can see that I'm explicitly calling `IncludeScopes = true` on the new up of the ApplicationInsightsLoggerOptions object even though the Application Insight documentation says it's switched on by default.

Here's some application code dealing with logging a trace:

```
try
{
	using (_log.BeginScope("", new Dictionary<string, object>
	{
		{"messageId", message.MessageId},
		{"some other key", "some other value"}

	}))
	{
		
		_log.LogTrace(
			$"Azure Function successfully triggered for a message with a MessageId of '{message.MessageId}' "
			+ $"via the '{_configuration.GetEntry(ServiceBusConfigurationKeys.SubscriptionName)}' subscription on the '{_configuration.GetEntry(ServiceBusConfigurationKeys.TopicName)}' topic.");

		await MessageHandler.HandleMessage(message, topicCollector, cancellationToken);

	}

}
catch (Exception e)
{
	_log.LogError(e.Message);
	throw;
}
```

Everything seems to be working fine except that I'm just not getting the 'messageId' / 'some other key' properties in Application Insights:

![image](https://user-images.githubusercontent.com/3339340/95809259-1a09ce00-0d5a-11eb-94c8-d48c5a8233d3.png)

Any assistance working out what is wrong would be greatly appreciated, thanks.

Regards,

Paul

**Comment by cijothomas** [comment]
Would you be able to provide a minimal repro app showing this behaviour? The sample usage of scopes+logger.log should have resulted in telemetry containing the scope as custom properties - not easy to tell what's wrong with the overall setup without a minimal repro. The fact that you are running this inside functions *MAY* be an issue as Functions integration with ApplicationInsight is somewhat different than regular usage.

**Comment by cijothomas** [comment]
btw, you don't need scopes to do structured logging. I was thinking the following is the right way to achieve structured logging with ilogger:
```
log.LogTrace("Azure Function successfully triggered for a message with MessageId: {MessageId}", message.MessageId);
```
The above should result in customProperty "MessageId", "actual messageid" to the TraceTelemetry.

**Comment by CodeRedPaul** [comment]
Hi @cijothomas 

Thanks for that - I'll take a look at creating a mini repo and referencing it.

> btw, you don't need scopes to do structured logging. I was thinking the following is the right way to achieve structured logging with ilogger:
> 
> ```
> log.LogTrace("Azure Function successfully triggered for a message with MessageId: {MessageId}", message.MessageId);
> ```
> 
> The above should result in customProperty "MessageId", "actual messageid" to the TraceTelemetry.

I didn't know that a logging a message would create structured data however, the notion of scope is still really useful to me as I have a high-level service helper where I'd like to set scope items such as MessageId, CorrellationId, BusinessEvent, etc and then have that propagate down to all logged traces within that scope.

I'll get back shortly with a reference to the repo...

Cheers,

Paul

**Comment by CodeRedPaul** [comment]
HI @cijothomas 

Quick question - when hosting in Visual Studio, for this example to run, we need at a minimum to have:

- Application Insights Instance
- Service Bus Topic and Subscription (the Azure Function is wired up as a Service Bus Subscription trigger)

If I'm going to create a mini repo, is it acceptable for me to leave the application settings unpopulated for these and let folks create their own instances? I'm unsure how to create these in my own Azure subscription and grant perms on them for people outside my domain.

Cheers,

Paul

**Comment by CodeRedPaul** [comment]
Hi @cijothomas 

Actually I just found the issue - I had a colleague get back to me and stepping through the code together we found the culprit. it was my bad...

The offending line was `_log.BeginScope("", new Dictionary<string, object>...`

This signature used the "Microsoft.Extensions.Logging.LoggerExtensions.BeginScope" extension method which takes in an initial parameter of `string messageFormat`.

If I remove the empty string, thus `_log.BeginScope(new Dictionary<string, object>`, it uses the `BeginScope<TState>(TState state)` method on 'Microsoft.Extensions.Logging.ILogger' and successfully adds the defined scope dictionary to Custom Properties as expected:

![image](https://user-images.githubusercontent.com/3339340/95829531-0a05e480-0d82-11eb-9c31-381b7f86d88a.png)

The screenshot above actually shows a trace deeper down the application's call stack (within `await MessageHandler.HandleMessage(message, topicCollector, cancellationToken);`), proving that the scope is in fact propagated down to any logging calls made with the ILogger and within the defined scope.

Thanks for the assist anyway. :)

Cheers,

Paul
